lib/ui/channel_buffers.dart

```diff
@@ -0,0 +1,125 @@
+// found in the LICENSE file.
+
+part of dart.ui;
+
+/// A saved platform message for a channel with its callback.
+class StoredMessage {
+  /// Default constructor, takes in a [ByteData] that represents the
+  /// payload of the message and a [PlatformMessageResponseCallback]
+  /// that represents the callback that will be called when the message
+  /// is handled.
+  StoredMessage(this._data, this._callback);
+
+  final ByteData _data;
+  final PlatformMessageResponseCallback _callback;
+
+  ByteData get data => _data;
+  PlatformMessageResponseCallback get callback => _callback;
+}
+
+/// A fixed-size circular queue.
+class _RingBuffer<T> {
+  final collection.ListQueue<T> _queue;
+  int _capacity;
+
+  _RingBuffer(this._capacity)
+    : _queue = collection.ListQueue<T>(_capacity);
+
+  int get length => _queue.length;
+
+  int get capacity => _capacity;
+
+  bool get isEmpty => _queue.isEmpty;
+
+  /// Returns true on overflow.
+  bool push(T val) {
+    bool overflow = false;
+    while (_queue.length >= _capacity) {
+      _queue.removeFirst();
+      overflow = true;
+    }
+    _queue.addLast(val);
+    return overflow;
+  }
+
+  /// Returns null when empty.
+  T pop() {
+    return _queue.isEmpty ? null : _queue.removeFirst();
+  }
+
+  /// Returns the number of discarded items resulting from resize.
+  int resize(int newSize) {
+    int result = 0;
+
+    while (length > newSize) {
+      result += 1;
+      _queue.removeFirst();
+    }
+
+    _capacity = newSize;
+
+    return result;
+  }
+}
+
+/// Storage of channel messages until the channels are completely routed.
+///
+/// Each channel has a finite buffer capacity and in a FIFO manner messages will
+/// be deleted if the capacity is exceeded.  The intention is that these buffers
+/// will be drained once a callback is setup on the BinaryMessenger in the
+/// Flutter framework.
+class ChannelBuffers {
+  /// A somewhat arbitrary size that tries to balance handling typical
+  /// cases and not wasting memory.
+  static const int kDefaultBufferSize = 100;
+
+  final Map<String, _RingBuffer<StoredMessage>> _messages =
+    <String, _RingBuffer<StoredMessage>>{};
+
+  /// Returns true on overflow.
+  bool push(String channel, ByteData data, PlatformMessageResponseCallback callback) {
+    _RingBuffer<StoredMessage> queue = _messages[channel];
+    if (queue == null) {
+      queue = _RingBuffer<StoredMessage>(kDefaultBufferSize);
+      _messages[channel] = queue;
+    }
+    final bool result = queue.push(StoredMessage(data, callback));
+    if (result) {
+      _Logger._printString('Overflow on channel: $channel.  '
+                           'Messages on this channel are being sent faster '
+                           'than they are being processed which is resulting '
+                           'in the dropping of messages.  The engine may not be '
+                           'running or you need to adjust the buffer size.');
```

Comment:
please only print to the console if it's actionable. Every message to the console should be an error that can be fixed in every circumstance.

Reply:
This is actionable and addressable as noted in the error message.  It means they are talking profusely to a dead channel.  They need to adjust the buffer size or run engine.

Reply:
@dnfield That sounds good to me, I've added a todo once resizing has percolated to the user level.

Reply:
We should either resolve this in this PR or remove the log IMO - I think this is probably really useful right now for someone doing local engine development, but it will likely cause confusion until it's actionable (imagine, a framework developer reading a bug report about this and not being familiar with this change).

Reply:
I prefer being more transparent in the interim, @xster what's your opinion?